### ISSUE TITLE
Document how eslint-plugin-jsdoc is different from valid-jsdoc

### ISSUE THREAD

**Issue by gajus** [gajus/eslint-plugin-jsdoc#1]
https://github.com/eslint/eslint/blob/master/docs/rules/valid-jsdoc.md

**Comment by khrome83** [comment]
There is a little overlap, but `eslint-plugin-jsdoc` does more with styling. 

From - https://github.com/eslint/eslint/blob/master/docs/rules/valid-jsdoc.md
> There is a JSDoc syntax error
2.A  @param  or  @returns  is used without a type specified
3.A  @param  or  @returns  is used without a description
4.A comment for a function is missing  @returns 
5.A parameter has no associated  @param  in the JSDoc comment
6. @param s are out of order with named arguments

**Comment by sebastienbarre** [comment]
Would be interested to know about that more (the overlap), since I'm trying to pick one or the other...
Thanks

**Comment by gajus** [comment]
@sebastienbarre I suggest to disable `valid-jsdoc` (because the two overlap) and use `eslint-plugin-jsdoc` because it has more rules.

**Comment by sebastienbarre** [comment]
Thanks @gajus

**Comment by sebastienbarre** [comment]
I can't quite figure out how to achieve the same as `valid-jsdoc`'s behavior for `@returns`. Specifically:
```
 * @returns {void}
```
does not trigger an error for `valid-jsdoc` (as documented [here](https://github.com/eslint/eslint/issues/673#issuecomment-37452054) and [here](http://eslint.org/docs/rules/valid-jsdoc#rule-details)), but triggers one for `jsdoc/require-returns-description`. I want to require a description for any other type, but admittedly (or conveniently?) returning `void` should not warrant a description, right?

Did I miss something? Thanks.

**Comment by sebastienbarre** [comment]
Bonus. Would you happen to know why this code is not triggering an error? 
The entire JSDoc block is missing.
```
var promiseRejectError = function() {
  return Promise.reject(Error('boom'));
};
```
I've ESLint's `require-jsdoc` turned on.
Thanks!

**Comment by gajus** [comment]
> Bonus. Would you happen to know why this code is not triggering an error? 
> The entire JSDoc block is missing.
> ```
> var promiseRejectError = function() {
>   return Promise.reject(Error('boom'));
> };
> ```
> I've ESLint's `require-jsdoc` turned on.
> Thanks!> 

http://eslint.org/docs/rules/require-jsdoc does not support `FunctionExpression` node.

**Comment by gajus** [comment]
> I can't quite figure out how to achieve the same as `valid-jsdoc`'s behavior for `@returns`. Specifically:
> ```
>  * @returns {void}
> ```
> does not trigger an error for `valid-jsdoc` (as documented [here](https://github.com/eslint/eslint/issues/673#issuecomment-37452054) and [here](http://eslint.org/docs/rules/valid-jsdoc#rule-details)), but triggers one for `jsdoc/require-returns-description`. I want > to require a description for any other type, but admittedly (or conveniently?) returning `void` should not warrant a description, right?
> 
> Did I miss something? Thanks.> 

`void` is not a valid JSDoc type. https://github.com/gajus/eslint-plugin-jsdoc/blob/master/README.md#eslint-plugin-jsdoc-rules-check-types

**Comment by sebastienbarre** [comment]
Thanks @gajus. What would you recommend? These don't work either:
```
 * @returns {undefined}
 * @returns
```
They all trigger `jsdoc/require-returns-description` error. There really isn't anything to describe here. But I also want that rule to be triggered when there is something to describe (ala `valid-jsdoc`).
Thanks

**Comment by gajus** [comment]
What is the reason for having returns tag when you are not returning anything?

**Comment by sebastienbarre** [comment]
My mistake, I thought I had read it was a standard practice, but it appears I was wrong. Apologies for the noise.

**Comment by gajus** [comment]
> My mistake, I thought I had read it was a standard practice, but it appears I was wrong. Apologies for the noise.

I have stopped using JSDoc type annotations in general. Switched to [Flow type](http://flowtype.org/) (without actually using flow). Having types in the code makes more sense than in a comment block. 

In case you decide to try it, there ESLint plugin for Flowtype ://github.com/gajus/eslint-plugin-flowtype and https://github.com/codemix/babel-plugin-typecheck can be used to inline conditions for type assertion (useful when running test suits)

**Comment by sebastienbarre** [comment]
Thanks @gajus; I had tried flow a while ago, but it wasn't very friendly to React.js code. Now it seems you can use flow with ESLint & Babel without having flow performing the syntax checks in the background? I'll give it a shot at some point, thanks for the pointers.

**Comment by gajus** [comment]
> Thanks @gajus; I had tried flow a while ago, but it wasn't very friendly to React.js code.

Use https://www.npmjs.com/package/babel-plugin-transform-flow-strip-types. This way it has zero effect to your React or whatever code.

**Comment by sebastienbarre** [comment]
Thanks. BTW I just updated, and had to revert back to 2.2.4. The new 2.3.0 is breaking in pretty much all of my files, on the first line:
```
lib/cli.js
  1:1  warning  Definition for rule 'jsdoc/require-hyphen-before-description' was not found  jsdoc/require-hyphen-before-description
```
Where the first line is just:
```
'use strict';
```

**Comment by gajus** [comment]
Sorry, I broke semver. Please replace `jsdoc/require-hyphen-before-description` rule with `jsdoc/require-hyphen-before-param-description`.

**Comment by ExE-Boss** [comment]
Since `valid‑jsdoc` has [been deprecated](https://eslint.org/blog/2018/11/jsdoc-end-of-life), this is&nbsp;now more important than ever.

**Comment by brettz9** [comment]
I've gotten something going now for this on the wiki: https://github.com/gajus/eslint-plugin-jsdoc/wiki/Comparison-with-deprecated-JSdoc-related-ESLint-rules

**Comment by brettz9** [comment]
In having done a review, the only feature I can see (besides any subtle differences that may exist between ESLint's deprecated `doctrine` project and jsdoctypeparser) that we don't have yet--and are therefore missing in the wiki docs--is our lack of `preferType`. But since this is tracked in #107 in terms of the `preferType` feature and need to update the wiki with documentation, I think this issue can be closed if there is nothing you see missing on the wiki page, @gajus ? https://github.com/gajus/eslint-plugin-jsdoc/wiki/Comparison-with-deprecated-JSdoc-related-ESLint-rules

**Comment by brettz9** [comment]
I updated the wiki to cross-reference what was the last remaining item from `valid-jsdoc`, its `preferType` option--now fulfilled by [`settings.jsdoc.preferredTypes`](https://github.com/gajus/eslint-plugin-jsdoc#settings-to-configure-check-types-and-no-undefined-types) (in conjunction with [`check-types`](https://github.com/gajus/eslint-plugin-jsdoc#options); [`no-undefined-types`](https://github.com/gajus/eslint-plugin-jsdoc#options-1) also has an option to use this map as well.)

Given your giving the thumbs up, I assume we are ok to close this now.
